### PR TITLE
Add `StreamMessage.endWith(finalizer)`

### LINKED ISSUE #4816: Conditionally emit the last message of a `StreamMessage`
As a future work of #4727, users might want to dynamically emit the last message depending on whether the upstream publisher completes successfully or exceptionally.
```java
interface StreamMessage<T> {

   <T> StreamMessage<T> endWith(Function<@Nullable Throwable, ? extends T> finalizer) {
       // Conditionally appends an element using the finalizer
       return new SurroundingPublisher(null, this, finalizer);
   }
}
```

### PR DESCRIPTION
**Pull request by injae-kim**
Related issue #4816 

users might want to dynamically emit the last message depending on whether the upstream publisher completes successfully or exceptionally.

- On #4816, it's good to add `StreamMessage.endWith(finalizer)`

### Modifications:

- Add `StreamMessage.endWith(finalizer)`

### Result:

- Closes #4816 
- Now user can dynamically emit the last value depending on whether the `StreamMessage` completes successfully or exceptionally by using `StreamMessage.endWith(finalizer)`

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 1]
```diff
@@ -366,7 +396,12 @@ public void onNext(T item) {
         @Override
         public void onError(Throwable cause) {
             requireNonNull(cause, "cause");
-            close(cause);
+            if (finalizer != null) {
+                downstream.onNext(finalizer.apply(cause));
+                close0(null);
+            } else {
+                close(cause);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:403]: I think we can also call `close0(cause)` right away.

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java [hunk 2]
```diff
@@ -1073,4 +1074,33 @@ default InputStream toInputStream(Function<? super T, ? extends HttpData> httpDa
         requireNonNull(executor, "executor");
         return new StreamMessageInputStream<>(this, httpDataConverter, executor);
     }
+
+    /**
+     * Emits the last value depending on whether this {@link StreamMessage}
+     * completes successfully or exceptionally.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:1080]: I think it doesn't emit the value depending on whether it is completed successfully or exceptionally. It always emits a value. 😉 
How about removing `depending on`?

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 3]
```diff
@@ -67,6 +75,22 @@ public SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher,
             this.publisher = new PublisherBasedStreamMessage<>(publisher);
         }
         this.tail = tail;
+        finalizer = null;
+    }
+
+    @SuppressWarnings("unchecked")
+    private SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher,
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:82]: nit: If you want, you can define another constructor that will dedup the assigning code.

**Review by minwoox — APPROVED** [review]
Looks great! Thanks!

**Review by jrhee17 — APPROVED** [review]
Looks good once @minwoox 's comments are addressed. Thanks! 🙇 👍 🙇

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 1 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java]: ✅ updated!

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java (hunk 2 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java]: ✅ updated!

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 3 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:82]: image

```java
public SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher,
                            Function<@Nullable Throwable, ? extends T> finalizer) { .. }

new SurroundingPublisher<>(null, Mono.just(1), null /*tail or finalizer?*/),
```
oh If we add above constructor, my IDE can't select the proper constructor on above `tail: null` case and test failed 😅

Is there any other good way?

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 4]
```diff
@@ -67,6 +75,22 @@ public SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher,
             this.publisher = new PublisherBasedStreamMessage<>(publisher);
         }
         this.tail = tail;
+        finalizer = null;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:78]: Should we convert `tail` to `finalizer` and delegate instead?
```java
static HttpResponse of(ResponseHeaders headers, 
                       Publisher<? extends HttpData> publisher, 
                       HttpHeaders trailers) {
    return of(headers, publisher, ignored -> trailers);
}

static HttpResponse of(ResponseHeaders headers, 
                       Publisher<? extends HttpData> publisher, 
                       Function<@Nullable Throwable, HttpHeaders> trailersFunction) { ... }
```

Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:78]: oh it's nice! ✅ updated.

If we use `tailersFunction: ignored -> trailers` on `HttpResponse.of`, `trailers` is emitted when publisher error. is it okay?

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 3 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:82]: Doesn't this work? 😄 
```java
public SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher, @Nullable T tail) {
    this(head, publisher, tail, null);
}

private SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher,
                             Function<@Nullable Throwable, ? extends T> finalizer) {
    this(head, publisher, null, requireNonNull(finalizer, "finalizer"));
}

private SurroundingPublisher(
        @Nullable T head, Publisher<? extends T> publisher,
        @Nullable T tail, @Nullable Function<@Nullable Throwable, ? extends T> finalizer) {
    requireNonNull(publisher, "publisher");
    this.head = head;
    if (publisher instanceof StreamMessage) {
        //noinspection unchecked
        this.publisher = (StreamMessage<T>) publisher;
    } else {
        this.publisher = new PublisherBasedStreamMessage<>(publisher);
    }
    this.tail = tail;
    this.finalizer = finalizer;
}
```

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 5]
```diff
@@ -59,14 +67,27 @@ public final class SurroundingPublisher<T> implements StreamMessage<T> {
 
     @SuppressWarnings("unchecked")
     public SurroundingPublisher(@Nullable T head, Publisher<? extends T> publisher, @Nullable T tail) {
+        this(head, publisher, tail, null);
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:73]: ditto.

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 6]
```diff
@@ -59,14 +67,27 @@ public final class SurroundingPublisher<T> implements StreamMessage<T> {
 
     @SuppressWarnings("unchecked")
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:68]: We can now remove this.

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 3 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:72]: oh~ this looks better. thanks! ✅ updated

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 6 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java]: ✅ updated!

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 7]
```diff
@@ -50,23 +56,36 @@ public final class SurroundingPublisher<T> implements StreamMessage<T> {
     private final StreamMessage<T> publisher;
     @Nullable
     private final T tail;
+    @Nullable
+    private final Function<@Nullable Throwable, ? extends T> finalizer;
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:60]: How about completely removing `tail` and using `finalizer` only?
If trailers are empty or null, we may set a special function such as 
```java
static final Function<@Nullable Throwable, ? extends T> nullTrailers = () -> null;

if (finalizer == nullTrailers) {
   // complete immediately. 
}
```

File: core/src/main/java/com/linecorp/armeria/common/PublisherBasedHttpResponse.java [hunk 8]
```diff
@@ -38,6 +40,12 @@ static PublisherBasedHttpResponse from(ResponseHeaders headers,
         return new PublisherBasedHttpResponse(new SurroundingPublisher<>(headers, publisher, trailers));
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/PublisherBasedHttpResponse.java:40]: This method no longer is used.

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 7 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:60]: > How about completely removing tail and using finalizer only?

aha you mean deleting constructor `new SurroundingPublisher("head", upstream, "tail")` too?

or just remove `tail` field and keep `new SurroundingPublisher("head", upstream, "tail")`? (replace "tail" to th -> "tail" finalizer internally)

Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:60]:
```
// T tail -> removed
Func<> finallizer
..

public onError() { // upstream error
  if (finalizer != null) {
     // If we use finalizer only, upstream error always replaced with item from finalizer.
    downstream.onNext(finalizer.apply(..));
  }
}
```
I worry that if we remove `tail` and use `finalizer` only, upstream error always replaced with finalizer.
But on our current logic with `tail`, upstream error is propagated on onError. WDYT? 🤔

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:60]: Good point. How about allowing `finalizer` not to handle the cause?
If `finalizer.apply(cause)` returns null, we can assume users do not want to recover the exception and finish the stream exceptionally.

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java [hunk 9]
```diff
@@ -1073,4 +1074,32 @@ default InputStream toInputStream(Function<? super T, ? extends HttpData> httpDa
         requireNonNull(executor, "executor");
         return new StreamMessageInputStream<>(this, httpDataConverter, executor);
     }
+
+    /**
+     * Emits the last value whether this {@link StreamMessage} completes successfully or exceptionally.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:1079]:
```suggestion
     * Emits the last value depending on whether this {@link StreamMessage} completes successfully or exceptionally.
```

**Review by trustin — COMMENTED** [review]
What would be the expected behavior when `finalizer.apply()` throws an exception? Could you make sure that the stream is closed even if a finalizer throws an exception so that the stream is not stalled? (+ test case please)

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 7 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java]: I applied this change myself. PTAL. 🙇‍♂️

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java [hunk 10]
```diff
@@ -321,14 +297,25 @@ private void sendHead() {
...
-                this.tail = null;
+            finalize(null);
+        }
+
+        private void finalize(@Nullable Throwable cause) {
+            final T tail;
+            try {
+                tail = finalizer.apply(cause);
+            } catch (Throwable ex) {
+                close0(ex);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:308]: How about leaving a log and close with the original `cause`?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java:308]: Good idea.

File: core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java (hunk 7 above)
Comment (injae-kim) [reply core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java]: oh sorry for late check, thanks for your update! 🙇

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java [hunk 11]
```diff
@@ -1073,4 +1074,37 @@ default InputStream toInputStream(Function<? super T, ? extends HttpData> httpDa
...
+     *          return -1;
+     *     });
+     * List<Integer> collected = endWith.collect().join();
+     *
+     * assert collected.equals(List.of(1, 2, 3, 4, 5, 100));
+     * }</pre>
+     *
+     * <p>Note that if {@code null} is returned by the {@link Function}, the {@link StreamMessage} will complete
+     * successfully without emitting an additional value when this stream is complete successfully,
+     * or complete exceptionally when this stream complete exceptionally.
```
Comment (injae-kim) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:1104]:
```java
    @Test
    void finalizer_return_null_upstream_thrown() {
        // given
        final StreamMessage<Integer> streamMessage = StreamMessage.of(2, 3, 4, 5);
        final StreamMessage<Integer> aborted = streamMessage
                .peek(val -> {
                    if (val == 5) {
                        throw new RuntimeException();
                    }
                });
        final Function<Throwable, Integer> finalizer = unused -> null;
        final StreamMessage<Integer> publisher = new SurroundingPublisher<>(1, aborted, finalizer);

        // when & then
        StepVerifier.create(publisher)
                    .expectNext(1, 2, 3, 4)
                    .expectError(RuntimeException.class) // 👈 👈 upstreamException is propagated when finalizer return null
                    .verify();
    }
```
@ikhoon , If `finalizer` returns `null` and upstream thrown, `finalize()` propagate exception to downstream with `close(upstreamException)`.

So I updated the comment in more detail~!

**Review by ikhoon — APPROVED** [review]
Thanks, @injae-kim! ❤️❤️

**Comment by minwoox** [comment]
Thanks a lot for adding this nice feature, @injae-kim! 🚀 🚀 🚀

**Pull request merged**